balancer/pickfirstleaf/pickfirstleaf.go

```diff
@@ -0,0 +1,658 @@
+// becomes ready, which means that all other subConn must be shutdown.
+// Only executed in the context of a serializer callback.
+func (b *pickfirstBalancer) shutdownRemaining(selected *scData) {
+	for _, v := range b.subConns.Values() {
+		sd := v.(*scData)
+		if sd.subConn != selected.subConn {
+			sd.subConn.Shutdown()
+		}
+	}
+	for _, k := range b.subConns.Keys() {
+		b.subConns.Delete(k)
+	}
+	b.subConns.Set(selected.addr, selected)
+}
+
+// requestConnection starts connecting on the subchannel corresponding to the
+// current address. If no subchannel exists, one is created. If the current
+// subchannel is in TransientFailure, a connection to the next address is
+// attempted.
+// Only executed in the context of a serializer callback.
+func (b *pickfirstBalancer) requestConnection() {
+	if !b.addressList.isValid() || b.state == connectivity.Shutdown {
+		return
+	}
+	curAddr := b.addressList.currentAddress()
+	sd, ok := b.subConns.Get(curAddr)
+	if !ok {
+		var err error
+		// We want to assign the new scData to sd from the outer scope, hence
+		// we can't use := below.
+		sd, err = b.newSCData(curAddr)
+		if err != nil {
+			// This should never happen, unless the clientConn is being shut
+			// down.
+			b.logger.Warningf("Failed to create a subConn for address %v: %v", curAddr.String(), err)
+			// The LB policy remains in TRANSIENT_FAILURE until a new resolver
+			// update is received.
+			b.state = connectivity.TransientFailure
+			b.addressList.reset()
+			b.cc.UpdateState(balancer.State{
+				ConnectivityState: connectivity.TransientFailure,
+				Picker:            &picker{err: fmt.Errorf("failed to create a new subConn: %v", err)},
+			})
+			return
+		}
+		b.subConns.Set(curAddr, sd)
+	}
+
+	scd := sd.(*scData)
+	switch scd.state {
+	case connectivity.Idle:
+		scd.subConn.Connect()
+	case connectivity.TransientFailure:
+		if !b.addressList.increment() {
+			b.endFirstPass(scd.lastErr)
+			return
+		}
+		b.requestConnection()
+	case connectivity.Ready:
+		// Should never happen.
+		b.logger.Errorf("Requesting a connection even though we have a READY subconn")
+	}
+}
+
+// Only executed in the context of a serializer callback.
+func (b *pickfirstBalancer) updateSubConnState(sd *scData, newState balancer.SubConnState) {
+	// Previously relevant subconns can still callback with state updates.
+	// To prevent pickers from returning these obsolete subconns, this logic
+	// is included to check if the current list of active subconns includes this
+	// subconn.
+	if activeSD, found := b.subConns.Get(sd.addr); !found || activeSD != sd {
+		return
+	}
+	if newState.ConnectivityState == connectivity.Shutdown {
+		return
+	}
+
+	if newState.ConnectivityState == connectivity.Ready {
+		b.shutdownRemaining(sd)
+		if !b.addressList.seekTo(sd.addr) {
+			// This should not fail as we should have only one subconn after
+			// entering READY. The subconn should be present in the addressList.
```

Comment:
This is _literally_ impossible because of `found` above, right?

So probably just call `seekTo` and ignore the return value, with a comment saying "it is impossible for seekTo to fail since we found the sc in the list above"?

Reply:
There are two data structures:
1. `b.subconns` which is a map from `resolver.Address->subConnData`.
2. `b.addressList`: A list of resolver.Address.

The implementation is responsible for ensuring that all the keys of `b.subconns` are present in the `addressList` by updating both of them as resolver updates come. I think it is possible that a bug in `reconcileSubConns()` causes an old subconn to not be removed from the map and we log an error here.

Reply:
OK that makes sense..  The invariant is between two data structures.